### DISCUSSION TITLE
Retro qns on username validation changes and patches made for related errors.

### DISCUSSION THREAD

**Discussion by RiyaMathew-11** [fossunited/fossunited#569] [Q&A]
Recently we got some issues where the user signup/sign-in was disrupted with a Timestamp error, which got me thinking on the process flow. 

In our hooks.py file, we have two methods described after a User signup occurs: 

> doc_events = {
>     "User": {
>         "after_insert": [
>             "fossunited.fossunited.user_utils.set_unique_username",
>             "fossunited.fossunited.user_utils.create_profile_on_user_create",
>         ],

Ideally both these methods should occur sequentially - first [`set_unique_username()`](https://github.com/fossunited/fossunited/blob/aa9ab87e114ba5a844979caf82d4ababc1fa5804/fossunited/fossunited/user_utils.py#L8), then [`create_profile_on_user_create()`](https://github.com/fossunited/fossunited/blob/aa9ab87e114ba5a844979caf82d4ababc1fa5804/fossunited/fossunited/user_utils.py#L16C5-L16C34). 

> ### Question 1: Is it guaranteed that when written in hooks, these methods are called sequentially, or is there a possibility of concurrent execution for the same after_insert event?

---------------------------

## Recent changes and errors

We made the [following changes](https://github.com/fossunited/fossunited/pull/541/files#diff-9bdb8414d304b86a3887f553b3d47da438adcca131f658be21d6784e917abb56) to both methods, in a recent attempt to strengthen username validations.

After changes:

```
def set_unique_username(doc, method):
    full_name = doc.full_name.lower()
    doc.first_name = full_name.split(" ")[0]
    doc.last_name = " ".join(full_name.split(" ")[1:])
    initial_username = re.sub(r"[^a-z0-9_]", "", full_name.replace(" ", "_"))
    doc.username = generate_username(initial_username)


def create_profile_on_user_create(doc, method):
    """
    Automatically Create a FOSS User Profile on User Creation / Signup
    """
    if not frappe.db.exists(
        USER_PROFILE,
        {"email": doc.email},
    ):
        profile = frappe.get_doc(
            {
                "doctype": USER_PROFILE,
                "user": doc.name,
                "full_name": doc.full_name,
                "username": doc.username,
                "is_published": 1,
            }
        )
        profile.insert(ignore_permissions=True)

    try:
        frappe.db.set_value("User", profile.user, "username", profile.username)
    except Exception as e:
        frappe.throw("Error updating username. Error: " + str(e))
```

Post-changes, we encountered [Timestamp errors](https://github.com/fossunited/fossunited/issues/558) during user signup during user signup and OAuth/social login. Adding `update_modified=False` in the `set_value()` method seemed to resolve the issue:

    try:
        frappe.db.set_value("User", profile.user, "username", profile.username, update_modified=False)
    except Exception as e:
        frappe.throw("Error updating username. Error: " + str(e))

After the patch, we thought on how the situation came in, for which we came up with hypotheses. 

## @harshtandiya's Thoughts:

-  We were not saving the User document in `set_unique_username()` where the username was generated initially. 
-  While on the second event -  the username is being changed with `set_value()` (which is more of a db call). (This was written to make sure the same username exists in the User and FOSS User Profile) 

The current patch of adding `update_modified = False` flag (updates without updating the `modified` timestamp) - hence removing timestamp errors on User doctype.

- The right way would be to save the doc in `set_unique_username()`, to safely remove the usage of `set_value()`. Such that there are no interferences/triggers on the doctype at the same time.

-------------

## My thoughts:

Let's check the entire sign-up flow. After a document is created (after_insert) in User doctype:

1. `set_unique_username()` is executed - username is created, but not saved.

   - There is no apparent change in the user doctype here but a doc event is triggered

2. `create_profile_on_user_create()` is executed: 

   - If the username is not saved in the other method: 

> ### Question 2: From where is the username fetched/which username is fetched for creating a profile ? 

  - Let's just say it got fetched and user profile created and User doctype value for username got set with a `frappe.db.set_value()` call. 

 - That should end the process there - both doctype methods got executed. 

 - Which further process is getting executed to get a timestamp error ? 

 **Unless - these methods occur concurrently.** 

----------

## Conclusion

This sort of clears the fact that: 

- There are 2 different doctype events here on the same doctype with it's mention in hooks. That brings me to the third question: 

> ### Question 3: Do both these methods have to be `doc_events` set in hooks?

- It's clear that both these methods are part of the same workflow, and is needed only once in a user's cycle. 

- [`set_unique_username()`](https://github.com/fossunited/fossunited/blob/aa9ab87e114ba5a844979caf82d4ababc1fa5804/fossunited/fossunited/user_utils.py#L8)'s purpose is to create a unique username, and could be called from [`create_profile_on_user_create()`](https://github.com/fossunited/fossunited/blob/aa9ab87e114ba5a844979caf82d4ababc1fa5804/fossunited/fossunited/user_utils.py#L16C5-L16C34)

- While `create_profile_on_user_create()` can be called in hooks. 

- This ensures that set_unique_username is not a `doc_event`, but together the entire flow creates a username and a FOSS profile, while also setting the same value of username to the User doctype.

**Comment by RiyaMathew-11** [comment]
## Some additional context: 

1. Frappe's core User doctype has the workflow which [sets a username on a new user](https://github.com/frappe/frappe/blob/93b1b1f0eac01bed3f6b99a46b2623b15cc5c12b/frappe/core/doctype/user/user.py#L679) sign-up. 

   - However this fails to generate a username for a user who has a similar firstname. 
   - It seems our `set_unique_username()` method was written to counter this fall. 

2. Checked some of the Frappe apps like LMS, HRMS, Insights, Erpnext etc - to see if functions with similar controller methods for doc_events  are mentioned in hooks.py. 

   - While [Erpnext](https://github.com/frappe/erpnext/blob/a1183f01652567dc4d7fdfa250dfd428cfe1dcda/erpnext/hooks.py#L313) has similar multiple functions written in same controller method for User doctype - all the methods mentioned seem to be mutually exclusive - unlike our case where `set_unique_username()` and `create_profile_on_user_create()` can be part of the same workflow.
  
   - Same case with [HRMS](https://github.com/frappe/hrms/blob/32ef661fdc8a1ed4844de59c8aa54c39a2fb60d0/hrms/hooks.py#L157)
 
   - While [LMS](https://github.com/frappe/lms/blob/e69cc9af1a5247f3c8f996b9370230cea401d9e3/lms/hooks.py#L99) do not have mutliple functions written in hooks for same controller methods in a doctype 

   - And [Insights](https://github.com/frappe/insights/blob/7b26aab941d9c4995ae1b211c6aa1a2890ea1129/insights/hooks.py#L123) don't seem to have any doc_events mentioned in hooks at all.

This kind of brings me back to question 3 in the above comment:

> ### Question 3: Do both these methods have to be doc_events set in hooks?

**Comment by shridarpatil** [comment]
Change `after_insert` to `before_insert` in hooks.py to set the user name in both cases. Also you don't need to explicitly set value using frappe.db.set_value you can just do `doc.username = <your user name>`.

For reference, you can check the robinhood army code where we have already implemented this to create a unique username `https://github.com/fossunited/robinhood/blob/master/robinhood/utils/user.py` 
also check hooks.py `https://github.com/fossunited/robinhood/blob/master/robinhood/hooks.py`


One more thing, Dont add `create_profile_on_user_create`  to hooks.py you can call this method from `set_unique_username` and create a user profile, there is no need to again set a username in this method.
